shared-bindings/camera/Camera.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright 2020 Sony Semiconductor Solutions Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/objproperty.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/camera/Camera.h"
+#include "shared-bindings/util.h"
+
+//| class Camera:
+//|     """The class to control camera.
+//|
+//|     Usage::
+//|
+//|         import board
+//|         import sdioio
+//|         import storage
+//|         import camera
+//|
+//|         sd = sdioio.SDCard(
+//|             clock=board.SDIO_CLOCK,
+//|             command=board.SDIO_COMMAND,
+//|             data=board.SDIO_DATA,
+//|             frequency=25000000)
+//|         vfs = storage.VfsFat(sd)
+//|         storage.mount(vfs, '/sd')
+//|
+//|         cam = camera.Camera(camera.ImageSize.IMAGE_SIZE_1920x1080)
+//|
+//|         file = open("/sd/image.jpg","wb")
+//|         cam.take_picture()
+//|         file.write(cam.picture)
+//|         file.close()"""
```

Comment:
I'd prefer passing in a file-like object or a bytearray. This could then be:
```suggestion
//|         with open("/sd/image.jpg","wb") as file:
//|             cam.take_picture(file)"""
```

It could also be:
```python
b = bytearray(width * height)
cam.take_picture(b)
```

What format will it produce? That seems missing in this API. I could see wanting a raw rgb565 to display on the screen for example.

Is it possible to change the image size on a per-picture basis?

Reply:
@tannewt I don't think we know the know the size of the buffer in advance if the image is compressed, so we can't necessarily pass it in. We know the max size but that may be far larger than the compressed size.

Reply:
If the low-level camera code could write directly to a stream (e.g. a file), then that would save having to allocate any RAM. But I don't know if that's possible.

Reply:
It i possible to change the image size on a per-picture basis. For example:
```
cam = camera.Camera(camera.ImageSize.IMAGE_SIZE_2048x1536)
cam.take_picture()

cam.size = camera.ImageSize.IMAGE_SIZE_1920x1080 
cam.take_picture()

cam.size = camera.ImageSize.IMAGE_SIZE_1280x960 
cam.take_picture()
```

Reply:
For the rest, please see my comment below.